src/backend/drm/egl/session.rs

```diff
@@ -15,44 +14,47 @@ use crate::backend::egl::{
     ffi,
     native::{Backend, NativeDisplay, NativeSurface},
 };
-use crate::backend::session::{AsSessionObserver, SessionObserver};
+use crate::{
+    backend::session::Signal as SessionSignal,
+    signaling::{Linkable, Signaler},
+};
 
 /// [`SessionObserver`](SessionObserver)
 /// linked to the [`EglDevice`](EglDevice) it was
 /// created from.
-pub struct EglDeviceObserver<S: SessionObserver + 'static, N: NativeSurface + Surface> {
-    observer: S,
+pub struct EglDeviceObserver<N: NativeSurface + Surface> {
     backends: Weak<RefCell<HashMap<crtc::Handle, Weak<EglSurfaceInternal<N>>>>>,
 }
 
-impl<S, B, D> AsSessionObserver<EglDeviceObserver<S, <D as Device>::Surface>> for EglDevice<B, D>
+impl<B, D> Linkable<SessionSignal> for EglDevice<B, D>
 where
-    S: SessionObserver + 'static,
     B: Backend<Surface = <D as Device>::Surface> + 'static,
     D: Device
         + NativeDisplay<
             B,
             Arguments = (crtc::Handle, Mode, Vec<connector::Handle>),
             Error = <<D as Device>::Surface as Surface>::Error,
-        > + AsSessionObserver<S>
+        > + Linkable<SessionSignal>
         + 'static,
     <D as Device>::Surface: NativeSurface,
 {
-    fn observer(&mut self) -> EglDeviceObserver<S, <D as Device>::Surface> {
-        EglDeviceObserver {
-            observer: self.dev.borrow_mut().observer(),
+    fn link(&mut self, signaler: Signaler<SessionSignal>) {
+        self.dev.borrow_mut().link(signaler.clone());
```

Comment:
What determines the order of the activate/pause calls of the different stacked devices, if we just pass down the signaler?

The order is important for some of the operations done in the different observers. I would prefer to call them explicitly.

Reply:
Callbacks are always invoked in the order in which they have been registered, so calling `link()` on the underlying device before doing anything ensures it'll always receive the notification before.